bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/handler/IpCameraHandler.java

```diff
@@ -0,0 +1,1999 @@
+import io.netty.handler.codec.http.FullHttpRequest;
+import io.netty.handler.codec.http.HttpClientCodec;
+import io.netty.handler.codec.http.HttpContent;
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.netty.handler.codec.http.HttpHeaderValues;
+import io.netty.handler.codec.http.HttpMessage;
+import io.netty.handler.codec.http.HttpMethod;
+import io.netty.handler.codec.http.HttpResponse;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.netty.handler.codec.http.HttpServerCodec;
+import io.netty.handler.codec.http.HttpVersion;
+import io.netty.handler.codec.http.LastHttpContent;
+import io.netty.handler.stream.ChunkedWriteHandler;
+import io.netty.handler.timeout.IdleState;
+import io.netty.handler.timeout.IdleStateEvent;
+import io.netty.handler.timeout.IdleStateHandler;
+import io.netty.util.CharsetUtil;
+import io.netty.util.ReferenceCountUtil;
+import io.netty.util.concurrent.GlobalEventExecutor;
+
+/**
+ * The {@link IpCameraHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Matthew Skinner - Initial contribution
+ */
+
+@NonNullByDefault
+public class IpCameraHandler extends BaseThingHandler {
+    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES = new HashSet<ThingTypeUID>(
+            Arrays.asList(THING_TYPE_ONVIF, THING_TYPE_HTTPONLY, THING_TYPE_AMCREST, THING_TYPE_DAHUA,
+                    THING_TYPE_INSTAR, THING_TYPE_FOSCAM, THING_TYPE_DOORBIRD, THING_TYPE_HIKVISION));
+    public static ArrayList<IpCameraHandler> listOfOnlineCameraHandlers = new ArrayList<IpCameraHandler>(1);
+    public static ArrayList<IpCameraGroupHandler> listOfGroupHandlers = new ArrayList<IpCameraGroupHandler>(0);
+    public static ArrayList<String> listOfOnlineCameraUID = new ArrayList<String>(1);
+    public final Logger logger = LoggerFactory.getLogger(getClass());
+    private ScheduledExecutorService cameraConnection = Executors.newScheduledThreadPool(1);
+    private ScheduledExecutorService scheduledMovePTZ = Executors.newScheduledThreadPool(1);
+    private ScheduledExecutorService pollCamera = Executors.newScheduledThreadPool(1);
+    private ScheduledExecutorService snapshot = Executors.newScheduledThreadPool(1);
+    public Configuration config;
+
+    // ChannelGroup is thread safe
+    public final ChannelGroup mjpegChannelGroup = new DefaultChannelGroup(GlobalEventExecutor.INSTANCE);
+    final ChannelGroup snapshotMjpegChannelGroup = new DefaultChannelGroup(GlobalEventExecutor.INSTANCE);
+    final ChannelGroup autoSnapshotMjpegChannelGroup = new DefaultChannelGroup(GlobalEventExecutor.INSTANCE);
+    public @Nullable Ffmpeg ffmpegHLS = null;
+    public @Nullable Ffmpeg ffmpegRecord = null;
+    public @Nullable Ffmpeg ffmpegGIF = null;
+    public @Nullable Ffmpeg ffmpegRtspHelper = null;
+    public @Nullable Ffmpeg ffmpegMjpeg = null;
+    public @Nullable Ffmpeg ffmpegSnapshot = null;
```

Comment:
Do you need a separate field for each of these? Isn't only one of them ever going to be non-null? Can you use a single field instead?

Reply:
Each one of those could be running in their own thread at the same time. So YES I do believe I need each of them ,unless I am not seeing something that you are because I am a self taught programmer with little experience.